### ISSUE TITLE
Convenience API for embedding model evaluation metrics as `ModelExplanation` element

### ISSUE THREAD

**Issue by dfarren** [jpmml/sklearn2pmml#410]
I would like to include model evaluation metrics in my PMML but it seems like [PMML Model Explanation](https://dmg.org/pmml/v4-3/ModelExplanation.html) elements are not supported in this project.

However, [I saw in the jpmml-sklearn tests that PMMLpipeline can accept custom pmml_options through pipeline.configure](https://github.com/jpmml/jpmml-sklearn/blob/1.5.9/src/test/resources/main.py#L563-L566). I assumed that this would include custom tags in the PMML, so I could include evaluation metrics as new tags.  This didn't work for me though.

I tried the snipped below but the PMML didn't change at all.
```
pmml_options["test_metric"] = {"test_key1": {"test_key2": 12345}}
pipeline.configure(**pmml_options)

```
How can I include metrics such as precision and recall in the PMML output?

**Comment by vruusmann** [comment]
> I would like to include model evaluation metrics in my PMML but it seems like [PMML Model Explanation](https://dmg.org/pmml/v4-4/ModelExplanation.html) elements are not supported in this project.

There is a related open issue in the JPMML-StatsModels project: https://github.com/jpmml/jpmml-statsmodels/issues/2

It was originally filed under the JPMML-SkLearn project, but I moved it over there. My reasoning was that Scikit-Learn - as a framework - does not care much about model explanation data, because this information is dataset dependent (ie. training vs validation vs testing), and currently it does not get collected or stored automatically during the `Estimator.fit(X, y)` method.

In contrast, StatsModels has it all available and embedded in the modeling results object automatically.

> However, [I saw in the jpmml-sklearn tests that PMMLpipeline can accept custom pmml_options through `pipeline.configure`. I assumed that this would include custom tags in the PMML.

The `Pipeline.configure(**pmml_options)` method is a convenience method for setting the `pmml_options_` attribute on the final estimator object.

The JPMML-SkLearn library uses these "hints" for choosing between alternate object representations. The prime usse case is choosing between "binary tree" and "multi-way split tree" representations for Scikit-Learn decision tree models (the latter being half the size of the former).

> How can I include metrics such as precision and recall in the PMML output?

There is currently no API (Python API at the SkLearn2PMML package level, or Java API at the underlying JPMML-SkLearn library level) for embedding arbitrary PMML/XML markup into PMML documents **as they are being generated**.

My recommended workaround would be this:
1. Use the SkLearn2PMML package to generate the PMML document.
2. Compute model evaluation metrics in Python.
3. Open the newly-generated PMML document in Python using whatever XML library you like the best, translate your model evaluation metrics into an appropriate `ModelExplanation` element, and insert it into the right place - manually.

The right place is typically the first top-level model element under the `/PMML/` root element. For example, for regression models this would be `/PMML/RegressionModel`.

**Comment by vruusmann** [comment]
This is a relevant issue, so keeping it open (but might move it around between JPMML projects again).

Basically, the `org.jpmml.converter.ModelEncoder` class (part of the JPMML-Converter library) should provide low-level API for attaching custom markup to the "main" model element when the PMML document is being generated.

The difficult part is designing a convenient high-level Python API for interacting with that low-level Java API method.

Should the end user be able to pass a fully formatted PMML/XML snippet there? In the beginning, could be passed as UTF-8 encoded string, and in future versions, as a Python [LXML](https://pypi.org/project/lxml/) library object.

Then again, I have my doubts that anyone would bother to prepare `ModelExplanation` elements manually in Python. So there should be an even higher-level API for generating LXML objects based on typical Scikit-Learn model evaluation metrics.

**Comment by vruusmann** [comment]
@dfarren Please share a concrete example about what kind of Scikit-Learn metrics you would like to embed into PMML.

Is it something trivial like storing the R^2 (in the form of [`sklearn.metrics.r2_score`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)), or is it something more complicated?

And who will be extracting and using this information in the end?

**Comment by vruusmann** [comment]
> I also include the threshold of the classifier used to reach their metrics (if it is a threshold-based classifier).

The threshold looks to be a model property (rather than a dataset property). 

So, strictly speaking, it's not a model evaluation metric per se?

>  Since a custom threshold is not supported by sklearn I think this is not a feature that would make sense to add to sklearn2pmml.

What kind of threshold-enabled classifier classes do you use? Could be some third-party library one.

I do recall that "thresholding" is very common with H2O.ai classifiers. But IIRC, the JPMML-H2O library (which is callable via the JPMML-SkLearn library) doesn't implement this part yet (due to no demand for it).

> This raises a new question: where in the PMML document would you advise me to insert the threshold used by the classifier?

Two options:

1. PMML lets you define extension elements. You can "extend" the main `ModelExplanation` element with custom XML markup if you please (eg. some company-specific dialect, for easy extraction by your model registry software).
2. As a classification model property. It would be rather straightforward to "convert" any  classification model into a thresholding-enabled classification model by creating a simple wrapper model element around it (takes the probability of the wrapped model, and then applies some thresholding or clipping function to it).

**Comment by vruusmann** [comment]
> It would be rather straightforward to "convert" any classification model into a thresholding-enabled classification model by creating a simple wrapper model element around it

Wouldn't be difficult to provide such a "thresholding" meta-classifier by the SkLearn2PMML package:

```python
classifier = ThresholdingBinaryClassifier(LogisticRegression(), threshold = 0.725)
```

**Comment by vruusmann** [comment]
> You can "extend" the main ModelExplanation element with custom XML markup if you please

My earlier suggestion about adding LXML library support would work very nicely with this!

Basically, you could define your own model evaluation metrics XML dialect - implemented as a reusable Python library - which would create "standardized" LXML objects for regression, classification etc. tasks.

Before conversion, you could then assign this LXML object to PMMLPipeline, and the JPMML-SkLearn converter would embed it into the PMML document automatically, under the `ModelExplanation` element of the main model.

**Comment by dfarren** [comment]
> The threshold looks to be a model property (rather than a dataset property).
>
> So, strictly speaking, it's not a model evaluation metric per se?

That's correct.

> What kind of threshold-enabled classifier classes do you use? Could be some third-party library one.
I am not aware of any threshold-enabled library TBH. I always end up adding this functionality myself.

**Comment by vruusmann** [comment]
> I am not aware of any threshold-enabled library TBH. I always end up adding this functionality myself.

OK, I might give the above `ThresholdingBinaryClassifier` meta-classifier idea a try in (one of-) the next SkLearn2PMML package versions.

This is only needed for binary classification tasks? Doesn't seem to generalize well for multi-class classification.

Care to share your thresholding post-processing code?

In H2O.ai it looks like below:
```java
if (_binomial) {
  preds[0] = (mu >= _defaultThreshold) ? 1 : 0; // threshold given by ROC
  preds[1] = 1.0 - mu; // class 0
  preds[2] =       mu; // class 1
} else {
  preds[0] = mu;
}
```

Looks like only the class label (`preds[0]`) is adjusted according to the threshold. The underlying probabilities (`preds[1]` and `preds[2]`) are still passed through unchanged.

**Comment by vruusmann** [comment]
> OK, I might give the above ThresholdingBinaryClassifier meta-classifier idea a try in (one of-) the next SkLearn2PMML package versions.

Thinking about it - a standalone `ThresholdingBinaryClassifier` class seems unnecessary, because it is possible to achieve similar H2O.ai-style thresholding using existing decision post-processing tools.

Here's an example:
```python
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn2pmml import sklearn2pmml
from sklearn2pmml.decoration import Alias
from sklearn2pmml.pipeline import PMMLPipeline
from sklearn2pmml.preprocessing import ExpressionTransformer

X, y = load_iris(return_X_y = True)
y = (y == 1).astype(int)

thresholder = Alias(ExpressionTransformer("1 if X[1] >= 0.75 else 0"), name = "threshold(y)", prefit = True)

pipeline = PMMLPipeline([
	("classifier", LogisticRegression())
], predict_proba_transformer = thresholder)
pipeline.fit(X, y)

sklearn2pmml(pipeline, "ThresholdingIris.pmml")
```

The resulting "ThresholdingIris.pmml" contains a model that predicts the original `y` target field, and `probability(0)` and `probability(1)` probability-type output fields. In addition, it predicts a `threshold(y)` output field, which is what we're talking about.

This example can be taken one step further, where the thresholding value becomes dynamic - can be passed in as a supplementary model input.

**Comment by vruusmann** [comment]
@dfarren I introduced the `PMMLPipeline.customize()` method, which lets you override the child elements of the main model element at the end of conversion (ie. no need for manual PMML document manipulation).

The usage is straightforward:
``` python
from lxml import etree

score = pipeline.score(X, y)

# All elements must specify PMML 4.4 XML namespace URI
model_explanation = etree.Element("{http://www.dmg.org/PMML-4_4}ModelExplanation")

predictive_model_quality = etree.SubElement(model_explanation, "{http://www.dmg.org/PMML-4_4}PredictiveModelQuality")
predictive_model_quality.attrib["targetField"] = ...
predictive_model_quality.attrib["r-squared"] = str(score)

# THIS!
pipeline.customize(etree.tostring(model_explanation))

sklearn2pmml(pipeline, ...)
```

Next up, the SkLearn2PMML package should provide utility functions for generating `PredictiveModelQuality` elements for most common regression- and classification-type tasks.

**Comment by vruusmann** [comment]
And when speaking about recording custom thresholds, then I'd currently suggest simply attaching a standard `Extension` child element with relevant information to a classification-type `PredictiveModelQuality` element:

```python
model_explanation = etree.Element("{http://www.dmg.org/PMML-4_4}ModelExplanation")

predictive_model_quality = etree.SubElement(model_explanation, "{http://www.dmg.org/PMML-4_4}PredictiveModelQuality")

# THIS!
extension = etree.SubElement(predictive_model_quality, "{http://www.dmg.org/PMML-4_4}Extension", name = "threshold", value = "0.75")

```

**Comment by vruusmann** [comment]
I'm trying to come up with an extra convenient API for embedding model metrics into PMML documents.

Now, the latest iteration is centered around `sklearn2pmml.metrics.PredictiveModelQuality` abstract base class (has `ClassifierQuality`, `BinaryClassifierQuality` and `RegressorQuality` concrete implementation classes):
https://github.com/jpmml/sklearn2pmml/blob/master/sklearn2pmml/metrics/__init__.py

Sample usage with a binary classification problem:
``` python
from sklearn2pmml.metrics import BinaryClassifierQuality

pipeline = PMMLPipeline([...])
pipeline.fit(X, y)

training_quality = BinaryClassifierQuality(estimator = pipeline, X = X_train, y = y_train, target_field = y.name, data_usage = "training") \
	.with_all_metrics()
test_quality = BinaryClassifierQuality(estimator = pipeline, X = X_test, y = y_test, target_field = y.name, data_usage = "test") \
	.with_all_metrics()

model_explanation = ModelExplanation() \
	.append(training_quality) \
	.append(test_quality)

# THIS!
pipeline.customize(model_explanation.tostring())

sklearn2pmml(pipeline, "pipeline.pmml")
```
